atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/runner/AntithesisWorkflowValidatorRunnerTest.java

```diff
@@ -0,0 +1,118 @@
+ */
+
+package com.palantir.atlasdb.workload.runner;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.palantir.atlasdb.workload.invariant.InvariantReporter;
+import com.palantir.atlasdb.workload.workflow.Workflow;
+import com.palantir.atlasdb.workload.workflow.WorkflowHistory;
+import com.palantir.atlasdb.workload.workflow.WorkflowValidator;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class AntithesisWorkflowValidatorRunnerTest {
+
+    private static final ListeningExecutorService EXECUTOR_SERVICE =
+            MoreExecutors.listeningDecorator(Executors.newFixedThreadPool(8));
+
+    @Mock
+    private Workflow workflow;
+
+    @Mock
+    private WorkflowHistory workflowHistory;
+
+    @Mock
+    private InvariantReporter<Void> invariantReporterOne;
+
+    @Mock
+    private InvariantReporter<Void> invariantReporterTwo;
+
+    private WorkflowValidator<Workflow> workflowValidator;
+
+    @Before
+    public void before() {
+        when(workflow.run()).thenReturn(workflowHistory);
+        workflowValidator = WorkflowValidator.of(workflow, invariantReporterOne, invariantReporterTwo);
+    }
+
+    @Test
+    public void runExecutesWorkflowAndInvokesInvariantReporters() {
+        new AntithesisWorkflowValidatorRunner(EXECUTOR_SERVICE).run(workflowValidator);
+        verify(workflow, times(1)).run();
+        verify(invariantReporterOne, times(1)).report(any());
+        verify(invariantReporterTwo, times(1)).report(any());
+    }
+
+    @Test
+    public void runValidatesAllInvariantsIgnoringExceptions() {
+        doThrow(new RuntimeException()).when(invariantReporterOne).report(any());
+        doThrow(new RuntimeException()).when(invariantReporterTwo).report(any());
+        new AntithesisWorkflowValidatorRunner(EXECUTOR_SERVICE).run(workflowValidator);
+        verify(workflow, times(1)).run();
+        verify(invariantReporterOne, times(1)).report(any());
+        verify(invariantReporterTwo, times(1)).report(any());
+    }
+
+    @Test
+    public void runExecutesMultipleWorkflowsAndWaitsForAllToFinishBeforeInvokingInvariantReporter() {
+        Semaphore semaphore = new Semaphore(0);
+        Workflow slowWorkflow = mock(Workflow.class);
+        when(slowWorkflow.run()).thenAnswer(_input -> {
```

Comment:
This is yucky, but I don't know think there's a better way to do this :/

Reply:
For the specification you describe, I think there's a nicer way. I'll send a small PR in :)

Reply:
see #6534?

Reply:
works for me -- I considered a similar approach but was worried about the race of latch being count-downed before the threads were released, but like should be fine in practice